Sanlouise addressvalidation e2es

sanlouise opened the pull request:
## Description
We needed to add end-to-end tests for address validation in the user profile.

## Acceptance criteria
- [x] Add sufficient e2e coverage with Cypress

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

sanlouise committed:
Working commit

sanlouise committed:
Add e2es

sanlouise committed:
Remove unneeded json

sanlouise committed:
2 suggestionsworks

sanlouise committed:
Remove unneeded json

sanlouise committed:
Use findByLabelText

sanlouise commented on src/applications/personalization/profile/tests/e2e/address-validation/bad-unit.cypress.spec.js:
```diff
@@ -0,0 +1,45 @@
+      cy.findByLabelText(/City/i)
+        .clear()
+        .type('San Francisco');
+      cy.get('#root_stateCode').select('CA');
```
For whatever reason, it finds more than one 'State' label - so I won't be using `findByLabelText('State')` here

sanlouise commented:
@erikphansen It's now finally ready for review! 💯

erikphansen reviewed (commented):
Looking great! Added a few thoughts/ideas. Lemme know what you think. Looking at this does make me realize we can likely update our code to skip that second call to the address validation API after making a selection.

erikphansen commented on src/applications/personalization/profile/tests/e2e/address-validation/addressValidation.js:
```diff
@@ -0,0 +1,310 @@
+export const createAddressValidationResponse = type => {
+  if (type === 'valid-address') {
```
One test I'd really like to see if when we get back a single, low confidence (`confidenceScore` < 90) address to confirm that we do show the validation screen in that situation.

sanlouise replied:
Added! 💪

sanlouise commented on src/platform/user/profile/vap-svc/containers/AddressValidationView.jsx:
```diff
@@ -276,6 +280,7 @@ class AddressValidationView extends React.Component {
           {!isLoading && (
             <button
               type="button"
+              data-testid="update-button"
```
I will remove this `data-testid` all together, it was a leftover from my previous tries to make the tests pass 😄

erikphansen commented on src/applications/personalization/profile/tests/e2e/address-validation/military-address.cypress.spec.js:
```diff
@@ -0,0 +1,37 @@
+        .type('PSC 808 Box 37');
+      cy.get('#root_addressLine2').clear();
+
+      cy.findByLabelText(/City/i).select('FPO');
```
this test seems to fail here since there is no `City` field once you select the military address checkbox.

erikphansen commented on src/applications/personalization/profile/tests/e2e/address-validation/military-address.cypress.spec.js:
```diff
@@ -0,0 +1,37 @@
+    it('should successfully update on Desktop', () => {
+      setUp('military');
+
+      cy.get('[type="checkbox"]')
```
This could be something like
```
cy.findByRole('checkbox', {
        name: /I live on a.*military base/i,
      }).check();
```

sanlouise committed:
Initial PR updates

sanlouise commented on src/applications/personalization/profile/tests/e2e/address-validation/addressValidation.js:
```diff
@@ -0,0 +1,310 @@
+    };
+  }
+
+  if (type === 'valid-address') {
```
oops! my bad 😅

sanlouise committed:
Add axe checks

sanlouise commented on src/applications/personalization/profile/tests/e2e/address-validation/bad-unit.cypress.spec.js:
```diff
@@ -0,0 +1,45 @@
+  describe('when entering an address with a bad unit', () => {
+    it('should successfully update on Desktop', () => {
+      setUp('bad-unit');
+
```
Definitely! Good point - added!

sanlouise committed:
Update PR per comments

sanlouise commented on src/applications/personalization/profile/tests/e2e/address-validation/bad-unit.cypress.spec.js:
```diff
@@ -0,0 +1,45 @@
+      cy.wait('@getUser');
+
+      cy.findByText(/225 irving st, Unit A/i).should('exist');
+      cy.findByText(/San Francisco, CA 94122/i).should('exist');
```
Added! 💯

sanlouise commented on src/applications/personalization/profile/tests/e2e/address-validation/bad-unit.cypress.spec.js:
```diff
@@ -0,0 +1,45 @@
+      });
+      cy.wait('@saveAddress');
+      cy.wait('@finishedTransaction');
+      cy.wait('@getUser');
```
I removed the other wait's and now just wait for the `getUser` call.

erikphansen replied:
with the next line being `findByTextId('mailingAddress')` this `cy.wait` isn't needed anymore :)

sanlouise commented on src/applications/personalization/profile/tests/e2e/address-validation/bad-unit.cypress.spec.js:
```diff
@@ -0,0 +1,45 @@
+        force: true,
+      });
+
+      cy.wait('@validateAddress');
```
Removed!

sanlouise committed:
Add low confidence score test

sanlouise committed:
Add data-testid to mailingAddress

sanlouise committed:
Undo

sanlouise committed:
Undo

sanlouise committed:
Update

erikphansen commented on src/applications/personalization/profile/tests/e2e/address-validation/low-confidence.cypress.spec.js:
```diff
@@ -0,0 +1,48 @@
+import { setUp } from 'applications/personalization/profile/tests/e2e/address-validation/setup';
+
+describe('Personal and contact information', () => {
+  describe('when entering a low confidence address that needs confirmation', () => {
```
could this be changed to something like `when getting a single suggestion with a confidence score <90`...

erikphansen commented on src/applications/personalization/profile/tests/e2e/address-validation/low-confidence.cypress.spec.js:
```diff
@@ -0,0 +1,48 @@
+
+describe('Personal and contact information', () => {
+  describe('when entering a low confidence address that needs confirmation', () => {
+    it('should successfully update on Desktop', () => {
```
and this to something like `should show the validation view and successfully update`?

erikphansen commented on src/applications/personalization/profile/tests/e2e/address-validation/addressValidation.js:
```diff
@@ -0,0 +1,307 @@
+export const createAddressValidationResponse = type => {
+  if (type === 'low-confidence') {
+    return {
+      addresses: [
```
Just because I'm being picky, I'd like to use a real response that has a confidence score in the 80s. This is for one reason in particular: it's be super easy for us to adjust the confidence score threshold we use to determine if we show the validation view or not. It used to be if the score was >80, we didn't show the screen. Then we decided to raise that threshold to 90. So I'd like this test to make sure we don't regress without catching it. It's actually easy to get a confidence score in the 80s. Just enter a valid address _but_ get the zip code wrong. Here is an example response:

```
{
  "addresses": [
    {
      "address": {
        "addressLine1": "36320 Coronado Dr",
        "addressType": "DOMESTIC",
        "city": "Fremont",
        "countryName": "United States",
        "countryCodeIso3": "USA",
        "countyCode": "06001",
        "countyName": "Alameda",
        "stateCode": "CA",
        "zipCode": "94536",
        "zipCodeSuffix": "5537"
      },
      "addressMetaData": {
        "confidenceScore": 81.0,
        "addressType": "Domestic",
        "deliveryPointValidation": "CONFIRMED",
        "residentialDeliveryIndicator": "RESIDENTIAL"
      }
    }
  ],
  "validationKey": -1808739503
}
```

sanlouise committed:
Fix low confidence test

sanlouise committed:
Remove cypress json

erikphansen commented on src/applications/personalization/profile/tests/e2e/address-validation/two-suggestions.cypress.spec.js:
```diff
@@ -0,0 +1,50 @@
+    it('should successfully update on Desktop', () => {
+      setUp('two-suggestions');
+
+      cy.get('#root_addressLine1')
```
Since these tests are really all about making sure the front end handles different responses from the address validation API, we _could_ input the same address into the form for every single test.

sanlouise committed:
Remove getUser waits

sanlouise committed:
Remove Cypress logic from isVAProfileServiceConfigured

sanlouise committed:
Fix

sanlouise committed:
Remove isvapservceconfigured check

sanlouise committed:
Fix linting

sanlouise committed:
Add window.Cypress to is VAProfileServiceConfigured

sanlouise committed:
Fix

erikphansen committed:
Explicitly mock the VAOS test user's contact info (#15063)

Since we can no longer rely on the default contact info that `mapRawUserDataToState()` was injecting into Redux

sanlouise committed:
Fix profile cypress tests

sanlouise committed:
Remove redundant test

erikphansen committed:
Explicitly mock the user contact info for HCA questionnaire tests (#15064)

Since we can no longer rely on the default contact info that mapRawUserDataToState() was injecting into Redux

sanlouise commented on src/applications/healthcare/questionnaire/tests/e2e/00.questionnaire.loads.cypress.spec.js:
```diff
@@ -1,4 +1,4 @@
-import basicUser from './fixtures/users/user-basic.json';
+import basicUser from './fixtures/users/user-basic.js';
```
To reviewers who wonder why the imports on their tests have changed:

In `isVAProfileServiceConfigured` we now include `window.Cypress`. The effect of this is that we now lo longer rely on `mockContactInformation` being set in `src/platform/user/profile/utilities/index.js` for Cypress tests.

This means that we now need to set the `vet360ContactInformation` in the tests' user objects to the mock contact information in `platform/user/profile/vap-svc/util/local-vapsvc.js`.

sanlouise commented:
To reviewers who wonder why the imports on their tests have changed:

In `isVAProfileServiceConfigured` we now include `window.Cypress`. The effect of this is that we now no longer rely on `mockContactInformation` being set in `src/platform/user/profile/utilities/index.js` for Cypress tests.

This means that we now need to set the `vet360ContactInformation` in the affected tests' user objects to the mock contact information in `platform/user/profile/vap-svc/util/local-vapsvc.js`.

mdewey reviewed (approved):
LGTM!
